src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,1336 @@
+                avoided_pollutant_load_path,
+                gdal.GDT_Float32,
+                NODATA),
+            target_path_list=[avoided_pollutant_load_path],
+            dependent_task_list=[retention_volume_task],
+            task_name=f'calculate avoided pollutant {pollutant} load'
+        )
+        aggregation_dependencies.append(avoided_load_task)
+        data_to_aggregate.append(
+            (avoided_pollutant_load_path, f'avoided_{pollutant}', 'sum'))
+
+    # (Optional) Do valuation if a replacement cost is defined
+    # you could theoretically have a cost of 0 which should be allowed
+    if (args['replacement_cost'] not in [None, '']):
+        valuation_task = task_graph.add_task(
+            func=pygeoprocessing.raster_calculator,
+            args=(
+                [
+                    (FILES['retention_volume_path'], 1), 
+                    (float(args['replacement_cost']), 'raw')
+                ],
+                retention_value_op, 
+                FILES['retention_value_path'], 
+                gdal.GDT_Float32, 
+                NODATA),
+            target_path_list=[FILES['retention_value_path']],
+            dependent_task_list=[retention_volume_task],
+            task_name='calculate stormwater retention value'
+        )
+        aggregation_dependencies.append(valuation_task)
+        data_to_aggregate.append(
+            (FILES['retention_value_path'], 'val_sum', 'sum'))
+        valuation_path = FILES['retention_value_path']
+    else:
+        valuation_path = None
+
+    # (Optional) Aggregate to watersheds if an aggregate vector is defined
+    if (args['aggregate_areas_path']):
+        reproject_aoi_task = task_graph.add_task(
+            func=pygeoprocessing.reproject_vector,
+            args=(
+                args['aggregate_areas_path'],
+                target_raster_info['projection_wkt'],
+                FILES['reprojected_aoi_path']),
+            kwargs={'driver_name': 'GPKG'},
+            target_path_list=[FILES['reprojected_aoi_path']],
+            task_name='reproject aggregate areas vector to match rasters',
+            dependent_task_list=[]
+        )
+        aggregation_dependencies.append(reproject_aoi_task)
+        aggregation_task = task_graph.add_task(
+            func=aggregate_results,
+            args=(
+                FILES['reprojected_aoi_path'],
+                data_to_aggregate),
+            target_path_list=[FILES['reprojected_aoi_path']],
+            dependent_task_list=aggregation_dependencies,
+            task_name='aggregate data over polygons'
+        )
+
+    task_graph.close()
+    task_graph.join()
+
+def threshold_array(array, threshold):
+    """Return a boolean array where 1 means less than the threshold value.
+    Assumes that the array's nodata value is the global NODATA.
+    Assume that nodata areas in the array are further than the threshold."""
+    out = numpy.full(array.shape, 0, dtype=numpy.int8)
```

Comment:
I chose `numpy.int8` on purpose to allow for the negative `NODATA`. I prefer the consistency that way over the small memory savings of using `numpy.uint8` and having to keep track of a particular nodata value like `127` just for this function. Happy to discuss more if you disagree!

Reply:
I've replaced the one `NODATA` with dtype-specific nodatas: `FLOAT_NODATA = -1`, `UINT8_NODATA = 255`, `UINT16_NODATA = 65535` and used the smaller data type where possible. This will definitely be an improvement for memory usage. I still like having them be global variables and not passing them in as function arguments (instead documenting this in the docstrings). E.g. `adjust_op` takes 4 arguments, and it would need 8 if I had to pass in a nodata value for each! What do you think?